### PR TITLE
Cache/Memory Management

### PR DESCRIPTION
**Pull request by acha-bill**
# Description
A cache layer that uses a FIFO eviction (release) policy and managed by a `cache manager`
Out of the box, the following viewmodels are cached.
 - TransactionViewModel
 - MilestoneViewModel
 - ApproveeViewModel
 
## Cache manager

Responsible for maintaining the caches. It is a singleton initialized by Guice.
It is initialized with a DbConfig that used used to configure the caches that it manages.
The cache manager is handled by the Tangle.

## Core operations
The cache maintains 2 concurrent hash maps. A strongstore and a weakstore. The strong store is built on strong values while the weak store is built using weak values.
When an item is to be released from cache, it is removed from the stongstore and put in the weakstore.
During a `get`, it checks the weakstore if the item is still around (if GC has not collected it) and returns it's value. Else, returns `null`.

### Release
An item is stored in DB immediately as it is also cached. This applies for TVM, AVM and MVM.
When a cached transaction is updated, it becomes `dirty`. 
Only dirty transactions will be saved to DB upon cache release.

## Cache Configuration
This feature introduces 4 new external config parameters.

- --tx-cache-size: The size of the tangle cache for transactions
- --tx-cache-release-count: The number of transactions to release from transaction cache
- --milestone-cache-size: The size of the tangle cache for milestones
- --milestone-cache-release-count: The number of milestones to release from the milestone cache

## Type of change

- Enhancement (a non-breaking change which adds functionality)

# How Has This Been Tested?

- New and existing unit tests pass
- New and existing regression tests pass

# Checklist:

- [x] My code follows the style guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/main/java/com/iota/iri/storage/Tangle.java [hunk 1]
```diff
@@ -206,7 +209,7 @@ private void updatePersistenceProvider(Persistable model, Indexable index, Strin
         }
     }
 
-    private void updateMessageQueueProvider(Persistable model, Indexable index, String item) {
+    public void updateMessageQueueProvider(Persistable model, Indexable index, String item) {
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/storage/Tangle.java]: Codacy Issue found: [Public method and constructor comments are required](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/test/java/com/iota/iri/service/snapshot/impl/SnapshotServiceImplTest.java [hunk 2]
```diff
@@ -10,20 +10,17 @@
...
-import org.junit.Assert;
-import org.junit.Before;
-import org.junit.FixMethodOrder;
-import org.junit.Rule;
-import org.junit.Test;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.junit.*;
```
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/service/snapshot/impl/SnapshotServiceImplTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/MainInjectionConfiguration.java [hunk 3]
```diff
@@ -35,22 +23,22 @@
...
-import com.iota.iri.service.tipselection.TailFinder;
-import com.iota.iri.service.tipselection.TipSelector;
-import com.iota.iri.service.tipselection.Walker;
-import com.iota.iri.service.tipselection.impl.CumulativeWeightCalculator;
-import com.iota.iri.service.tipselection.impl.EntryPointSelectorImpl;
-import com.iota.iri.service.tipselection.impl.TailFinderImpl;
-import com.iota.iri.service.tipselection.impl.TipSelectorImpl;
-import com.iota.iri.service.tipselection.impl.WalkerAlpha;
+import com.iota.iri.service.tipselection.*;
+import com.iota.iri.service.tipselection.impl.*;
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/MainInjectionConfiguration.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.service.tipselection.impl'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/test/java/com/iota/iri/service/milestone/impl/MilestoneServiceImplTest.java [hunk 4]
```diff
@@ -8,20 +8,16 @@
...
-import org.junit.After;
-import org.junit.Assert;
-import org.junit.Before;
-import org.junit.FixMethodOrder;
-import org.junit.Rule;
-import org.junit.Test;
+
+import java.util.Optional;
+
+import org.junit.*;
```
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/service/milestone/impl/MilestoneServiceImplTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/Iota.java [hunk 5]
```diff
@@ -21,14 +22,10 @@
 import com.iota.iri.service.transactionpruning.DepthPruningCondition;
 import com.iota.iri.service.transactionpruning.SizePruningCondition;
 import com.iota.iri.service.transactionpruning.TransactionPruner;
+import com.iota.iri.storage.*;
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/Iota.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.storage'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/MainInjectionConfiguration.java [hunk 6]
```diff
@@ -35,22 +23,22 @@
...
-import com.iota.iri.service.tipselection.RatingCalculator;
-import com.iota.iri.service.tipselection.TailFinder;
-import com.iota.iri.service.tipselection.TipSelector;
-import com.iota.iri.service.tipselection.Walker;
-import com.iota.iri.service.tipselection.impl.CumulativeWeightCalculator;
-import com.iota.iri.service.tipselection.impl.EntryPointSelectorImpl;
-import com.iota.iri.service.tipselection.impl.TailFinderImpl;
-import com.iota.iri.service.tipselection.impl.TipSelectorImpl;
-import com.iota.iri.service.tipselection.impl.WalkerAlpha;
+import com.iota.iri.service.tipselection.*;
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/MainInjectionConfiguration.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.service.tipselection'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/MainInjectionConfiguration.java [hunk 7]
```diff
@@ -15,16 +11,8 @@
...
-import com.iota.iri.service.milestone.MilestoneService;
-import com.iota.iri.service.milestone.MilestoneSolidifier;
-import com.iota.iri.service.milestone.SeenMilestonesRetriever;
-import com.iota.iri.service.milestone.impl.LatestMilestoneTrackerImpl;
-import com.iota.iri.service.milestone.impl.LatestSolidMilestoneTrackerImpl;
-import com.iota.iri.service.milestone.impl.MilestoneServiceImpl;
-import com.iota.iri.service.milestone.impl.MilestoneSolidifierImpl;
-import com.iota.iri.service.milestone.impl.SeenMilestonesRetrieverImpl;
+import com.iota.iri.service.milestone.*;
+import com.iota.iri.service.milestone.impl.*;
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/MainInjectionConfiguration.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.service.milestone.impl'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/MainInjectionConfiguration.java [hunk 8]
```diff
@@ -15,16 +11,8 @@
...
-import com.iota.iri.service.milestone.LatestSolidMilestoneTracker;
-import com.iota.iri.service.milestone.MilestoneService;
-import com.iota.iri.service.milestone.MilestoneSolidifier;
-import com.iota.iri.service.milestone.SeenMilestonesRetriever;
-import com.iota.iri.service.milestone.impl.LatestMilestoneTrackerImpl;
-import com.iota.iri.service.milestone.impl.LatestSolidMilestoneTrackerImpl;
-import com.iota.iri.service.milestone.impl.MilestoneServiceImpl;
-import com.iota.iri.service.milestone.impl.MilestoneSolidifierImpl;
-import com.iota.iri.service.milestone.impl.SeenMilestonesRetrieverImpl;
+import com.iota.iri.service.milestone.*;
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/MainInjectionConfiguration.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.service.milestone'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/storage/Tangle.java [hunk 9]
```diff
@@ -1,33 +1,27 @@
...
+import com.iota.iri.cache.Cache;
+import com.iota.iri.cache.CacheManager;
+import com.iota.iri.cache.impl.CacheManagerImpl;
+import com.iota.iri.conf.MainnetConfig;
+import com.iota.iri.controllers.ApproveeViewModel;
+import com.iota.iri.controllers.MilestoneViewModel;
+import com.iota.iri.controllers.TransactionViewModel;
+import com.iota.iri.model.Hash;
+import com.iota.iri.model.StateDiff;
+import com.iota.iri.model.persistables.*;
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/storage/Tangle.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.model.persistables'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/test/java/com/iota/iri/conf/ConfigTest.java [hunk 10]
```diff
@@ -13,25 +16,18 @@
...
 import org.apache.commons.lang3.StringUtils;
 import org.hamcrest.CoreMatchers;
 import org.hamcrest.core.IsCollectionContaining;
-import org.junit.After;
-import org.junit.AfterClass;
-import org.junit.Assert;
-import org.junit.BeforeClass;
-import org.junit.FixMethodOrder;
-import org.junit.Test;
+import org.junit.*;
```
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/conf/ConfigTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/test/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorImplTest.java [hunk 11]
```diff
@@ -10,11 +10,8 @@
 import com.iota.iri.service.snapshot.impl.SnapshotMockUtils;
 import com.iota.iri.service.tipselection.EntryPointSelector;
 import com.iota.iri.storage.Tangle;
-import org.junit.Assert;
-import org.junit.Before;
-import org.junit.FixMethodOrder;
-import org.junit.Rule;
-import org.junit.Test;
+
+import org.junit.*;
```
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorImplTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/controllers/MilestoneViewModel.java [hunk 12]
```diff
@@ -249,4 +227,69 @@ public void delete(Tangle tangle) throws Exception {
...
+                MilestoneViewModel milestoneViewModel = cache.get(index);
+                if (milestoneViewModel != null) {
+                    batch.add(new Pair<>(index, milestoneViewModel.milestone));
+                    indicesToRelease.add(index);
+                }
+            }
+        }
+
+        tangle.saveBatch(batch);
+        cache.release(indicesToRelease);
```
[the remark this answers is not in the document]
Comment (acha-bill) [reply src/main/java/com/iota/iri/controllers/MilestoneViewModel.java]: sure

**Comment by acha-bill** [comment]
Issue for tests: #1734 
Issue for cache stats: #1722

File: src/main/java/com/iota/iri/storage/Tangle.java (hunk 1 above)
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/storage/Tangle.java]: Codacy Issue found: [Public method and constructor comments are required](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/Iota.java (hunk 5 above)
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/Iota.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.storage'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/test/java/com/iota/iri/conf/ConfigTest.java (hunk 10 above)
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/conf/ConfigTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/storage/Tangle.java (hunk 9 above)
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/storage/Tangle.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.model.persistables'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/MainInjectionConfiguration.java (hunk 7 above)
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/MainInjectionConfiguration.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.service.milestone.impl'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/MainInjectionConfiguration.java (hunk 8 above)
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/MainInjectionConfiguration.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.service.milestone'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/test/java/com/iota/iri/service/snapshot/impl/SnapshotServiceImplTest.java (hunk 2 above)
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/service/snapshot/impl/SnapshotServiceImplTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/MainInjectionConfiguration.java (hunk 3 above)
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/MainInjectionConfiguration.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.service.tipselection.impl'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/test/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorImplTest.java (hunk 11 above)
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorImplTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/MainInjectionConfiguration.java (hunk 6 above)
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/MainInjectionConfiguration.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.service.tipselection'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/test/java/com/iota/iri/service/milestone/impl/MilestoneServiceImplTest.java (hunk 4 above)
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/service/milestone/impl/MilestoneServiceImplTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/storage/Tangle.java [hunk 13]
```diff
@@ -206,7 +205,7 @@ private void updatePersistenceProvider(Persistable model, Indexable index, Strin
         }
     }
 
-    private void updateMessageQueueProvider(Persistable model, Indexable index, String item) {
+    public void updateMessageQueueProvider(Persistable model, Indexable index, String item) {
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/storage/Tangle.java]: Codacy Issue found: [Public method and constructor comments are required](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/storage/Tangle.java [hunk 14]
```diff
@@ -1,33 +1,25 @@
...
-import java.util.function.Function;
-import java.util.stream.Collectors;
+import com.iota.iri.cache.Cache;
+import com.iota.iri.cache.CacheManager;
+import com.iota.iri.cache.impl.CacheManagerImpl;
+import com.iota.iri.conf.MainnetConfig;
+import com.iota.iri.controllers.TransactionViewModel;
+import com.iota.iri.model.Hash;
+import com.iota.iri.model.StateDiff;
+import com.iota.iri.model.persistables.*;
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/storage/Tangle.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.model.persistables'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/test/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorImplTest.java [hunk 15]
```diff
@@ -10,11 +9,8 @@
...
```
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorImplTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/MainInjectionConfiguration.java (hunk 7 above)
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/MainInjectionConfiguration.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.service.milestone.impl'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/MainInjectionConfiguration.java (hunk 8 above)
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/MainInjectionConfiguration.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.service.milestone'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/test/java/com/iota/iri/service/snapshot/impl/SnapshotServiceImplTest.java [hunk 16]
```diff
@@ -10,20 +9,17 @@
...
```
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/service/snapshot/impl/SnapshotServiceImplTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/MainInjectionConfiguration.java (hunk 6 above)
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/MainInjectionConfiguration.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.service.tipselection'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/test/java/com/iota/iri/service/milestone/impl/MilestoneServiceImplTest.java (hunk 4 above)
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/service/milestone/impl/MilestoneServiceImplTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/test/java/com/iota/iri/conf/ConfigTest.java (hunk 10 above)
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/conf/ConfigTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/Iota.java (hunk 5 above)
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/Iota.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.storage'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/test/java/com/iota/iri/service/milestone/impl/MilestoneServiceImplTest.java [hunk 17]
```diff
@@ -67,15 +63,13 @@ public void mockProcessed(Tangle tangle, boolean applied) {
...
+    public void setUp() throws Exception {
         SnapshotMockUtils.mockSnapshotProvider(snapshotProvider);
 
-        MilestoneViewModel.clear();
     }
 
     @After
-    public void tearDown() {
-        MilestoneViewModel.clear();
+    public void tearDown() throws Exception {
```
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/service/milestone/impl/MilestoneServiceImplTest.java]: Codacy Issue found: [Document empty method body](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

File: src/main/java/com/iota/iri/MainInjectionConfiguration.java (hunk 3 above)
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/MainInjectionConfiguration.java]: Codacy Issue found: [Avoid unused imports such as 'com.iota.iri.service.tipselection.impl'](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4896328)

**Pull request merged**